### PR TITLE
Fix upload security and working uploads when "upload_tmp_dir" is not included in "open_basedir"

### PR DESCRIPTION
**Pull request by mvorisek**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Fix upload security and working uploads when "upload_tmp_dir" is not included in "open_basedir". PHP file uploads are designed to upload files firstly to an secured location not accessible from PHP directly and then upload files are expected to be moved by `move_uploaded_file` method.<br>Currently PrestaShop has no centralized code uploads and their validation, there are a lot of usages (like checking if the file is really an image) before the `move_uploaded_file` method is called. This PR addresses all there issues by providing custom `is_uploaded_file` (`Uploader::isUploadedFile`) and `move_uploaded_file` (`Uploader::moveUploadedFile`) methods which behave like the PHP methods, do extra validation and move the uploaded files to temporaly location before they are finally moved. They also take care to unlink the temproraly files if not done so by the user/script itself.
| Type?         | bug fix / improvement
| Category?     | FO / BO (all)
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | no
| How to test?  | All tests should pass.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16643)

### CODE REVIEW

**Comment by PierreRambaud** [comment]
Since you're still working on it, I add the `[WIP]` in your pull request title.
Feel free to tell us when it's ok and what do you exactly change, because I see many lines not related to the upload_tmp_dir nor open_basedir.

**Comment by mvorisek** [comment]
@PierreRambaud It is complete now.

There are 3 commits:
- 1st one https://github.com/PrestaShop/PrestaShop/pull/16643/commits/648f656696db7d0b5dc7c47ad770b517dd23e2fa has self explanatory title
- 2nd one https://github.com/PrestaShop/PrestaShop/pull/16643/commits/34c75f2d601bbdfa931711a6a2139403f28a96cd resolves the wrong usage of `move_uploaded_file` thru the centralized & robust methods in the `Uploader` class plus:
    - `admin-dev/filemanager/upload.php` is rewritten extra, as it does not use PS autoloader
    - `Uploader::isUploadedFile` does a lot of additional checks, so some (and before in every place very different/incomplete) checks were removed, like https://github.com/PrestaShop/PrestaShop/pull/16643/commits/34c75f2d601bbdfa931711a6a2139403f28a96cd#diff-9079f61ba1c5da7f9f142723d5d42ea0L3919 or https://github.com/PrestaShop/PrestaShop/pull/16643/commits/34c75f2d601bbdfa931711a6a2139403f28a96cd#diff-c9209e1d590bac0281a599f2c5468a35L307
    - some hard-coded constant values were replaced with their constants names, like https://github.com/PrestaShop/PrestaShop/pull/16643/commits/34c75f2d601bbdfa931711a6a2139403f28a96cd#diff-af1f7ec6d2d688a167b76bfddee27afbL301
- 3rd one https://github.com/PrestaShop/PrestaShop/pull/16643/commits/1eef50b0f83fa220aa83b2a33a1416cb7e5c7f53 fixes additional issues with uploads related mainly with upload site limits. Ini values were parsed in many different ways, starting with (insane) removal of `M` char from the ini value in a hope that the value was defined with `M` multiplier like https://github.com/PrestaShop/PrestaShop/pull/16643/commits/1eef50b0f83fa220aa83b2a33a1416cb7e5c7f53#diff-38aac346959e2aa6a17c872bb71b5fddL83 or (not visible on the first sight here) https://github.com/PrestaShop/PrestaShop/pull/16643/commits/1eef50b0f83fa220aa83b2a33a1416cb7e5c7f53#diff-ea0413704cbbcf57a99ec2588856ac9cL254 . I have also carefully removed/refactored some code with duplicated code like https://github.com/PrestaShop/PrestaShop/pull/16643/commits/1eef50b0f83fa220aa83b2a33a1416cb7e5c7f53#diff-b5e854e1b6f2a41c568bff04fb910a9eL50 . Lastly I normalized all max. size limits to a central method `Tools::getMaxUploadSize()`, originally there was random usage between `upload_max_filesize` and `post_max_size` php.ini values with bad interpretation on some lines (`upload_max_filesize` limits the max. size per one file, but if `post_max_size` is smaller, it needs to be used as a limit) or bad conversions to bytes as described above.

If there is anything additional I can help with please let me know.

The code of interest for this issue/PR was found by these regexes:
```
\$_FILES.+?(?<!\$)tmp_name|(?<!\$)tmp_name
\$_?file.+?(?<!\$)(?<!Notifications\.)(?<!an )(?<!\w)error(?!'\]\[\])(?!'\]\s*=\s*arr)(?!'\]\s*=\s*\$this->trans)(?!'\]\s*=\s*(Tools|Context)::)
\$_FILES
```

File: admin-dev/filemanager/upload.php [hunk 1]
```diff
@@ -38,14 +38,14 @@
...
+if (!empty($_FILES['file']) && !empty($_FILES['file']['name']) && !empty($_FILES['file']['size'])
+        && is_uploaded_file($_FILES['file']['tmp_name'])) {
     $info = pathinfo($_FILES['file']['name']);
-    if (isset($info['extension'])
-            && in_array(fix_strtolower($info['extension']), $ext)
-            && in_array(mime_content_type($_FILES['file']['tmp_name']), $mime)
-    ) {
-        $tempFile = $_FILES['file']['tmp_name'];
-
+    $tempFile = sys_get_temp_dir() . '/ps__filemanager__' . time() . '__'
```
Comment (PierreRambaud) [inline admin-dev/filemanager/upload.php:44]: seems really overkill when php provides: https://www.php.net/manual/en/function.tmpfile.php
And Symfony also provides a way to do ti :thinking:

**Comment by PierreRambaud** [comment]
I just notice that this part of code came from https://github.com/trippo/ResponsiveFilemanager/tree/master/filemanager not sure it's a good idea to change the code.
We should use the repository instead :thinking:

File: admin-dev/filemanager/upload.php (hunk 1 above)
Comment (mvorisek) [reply admin-dev/filemanager/upload.php:44]: How can `tmpfile()` method be used with `move_uploaded_file`?

**Comment by mvorisek** [comment]
> 
> 
> I just notice that this part of code came from https://github.com/trippo/ResponsiveFilemanager/tree/master/filemanager not sure it's a good idea to change the code.
> We should use the repository instead 🤔

This code is not integrated as a submodule or composer package. I can propose the changes to the official repo too. Is this currently an issue for merging this PR? This is the only change of 3rd party code/code not using PrestaShop autoloader and mentioned it in the PR description.

Can you please also advide why the Travis CI job did not started after I force-pushed the changes advised form Pretty CI - can you rerun the tests?

**Comment by PierreRambaud** [comment]
@mvorisek We are planning to remove 3rd party code into composer / npm dependencies.
Btw, you should propose changes into the original repository :)

We need feedback for our lead dev @eternoendless before continue the review process :/

**Comment by mvorisek** [comment]
@PierreRambaud The current code of the `filemanager` is very different, compare https://github.com/trippo/ResponsiveFilemanager/blob/master/filemanager/upload.php . As this is the only modification of 3rd party code, I would advise merging this change too to fix the issue completely.

@eternoendless Please review, if there are any questions I am ready to answer them.

**Comment by mvorisek** [comment]
Can we proceed with this PR?

**Comment by mvorisek** [comment]
Some improvements can be cherrypicked, however I do not want to pursue the main issue as `filemanager` must be updated

**Pull request closed without merging**